File: java/src/main/java/ola/rpc/StreamRpcChannel.java
/***********************************************************************
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 *
 *************************************************************************/
package ola.rpc;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.logging.Level;
import java.util.logging.Logger;

import ola.proto.Ola.STREAMING_NO_RESPONSE;
import ola.rpc.Rpc.RpcMessage;

import com.google.protobuf.Descriptors.MethodDescriptor;
import com.google.protobuf.DynamicMessage;
import com.google.protobuf.Message;
import com.google.protobuf.RpcCallback;
import com.google.protobuf.RpcChannel;
import com.google.protobuf.RpcController;

/**
 * Basic RPC Channel implementation.  All calls are done
 * synchronously.
 *
 * The RPC Channel is hard coded to localhost 9010 where the
 * olad daemon is running.
 */
public class StreamRpcChannel implements RpcChannel {

    private static Logger logger = Logger.getLogger(StreamRpcChannel.class.getName());

    public static final int PORT = 9010;

    public static final String HOST = "localhost";

    private static final int PROTOCOL_VERSION = 1;

    private static final int VERSION_MASK = 0xf0000000;

    private static final int SIZE_MASK = 0x0fffffff;

    private Socket socket = null;

    private BufferedOutputStream bos;

    private BufferedInputStream bis;

    private int sequence = 0;


    /**
     * Create new Rpc Channel Connection to olad.
     * @throws Exception
     */
    public StreamRpcChannel() throws Exception {
        connect();
    }


    /**
     * Open connection to olad daemon.
     *
     * @throws Exception
     */
    public void connect() throws Exception {

        if (socket != null && socket.isConnected()) {
            logger.warning("Socket already connected.");
            return;
        }

        try {
            socket = new Socket(HOST, PORT);
            bos = new BufferedOutputStream(socket.getOutputStream());
            bis = new BufferedInputStream(socket.getInputStream());
        } catch (Exception e) {
            logger.severe("Error connecting. Make sure the olad daemon is running on port 9010");
            throw e;
        }
    }


    /**
     * Close Rpc Channel.
     */
    public void close() {

        if (socket != null && socket.isConnected()) {
            try {
                socket.close();
            } catch (Exception e) {
                logger.warning("Error closing socket. " + e.getMessage());
            }
        }
    }


    /* (non-Javadoc)
     * @see com.google.protobuf.RpcChannel#callMethod(com.google.protobuf.Descriptors.MethodDescriptor, com.google.protobuf.RpcController, com.google.protobuf.Message, com.google.protobuf.Message, com.google.protobuf.RpcCallback)
     */
    public void callMethod(MethodDescriptor method, RpcController controller,
            Message requestMessage, Message responseMessage, RpcCallback<Message> done) {

        int messageId = sequence++;

        RpcMessage message = RpcMessage.newBuilder()
                .setType(Rpc.Type.REQUEST)
                .setId(messageId)
                .setName(method.getName())
                .setBuffer(requestMessage.toByteString())
                .build();

        try {

            sendMessage(message);
            if (responseMessage.getDescriptorForType().getName().equals("STREAMING_NO_RESPONSE")) {
                // don't wait for response on streaming messages..
                return;
            }

            RpcMessage response = readMessage();

            if (response.getType().equals(Rpc.Type.RESPONSE)) {
                if (response.getId() != messageId) {
                    controller.setFailed("Received message with id " + response.getId() + " , but was expecting " + messageId);
                } else {
                    responseMessage = DynamicMessage.parseFrom(responseMessage.getDescriptorForType(), response.getBuffer());
                    if (done != null) {
                        done.run(responseMessage);
                    }
                }
            } else {
                controller.setFailed("No valid response received !");
            }


        } catch (Exception e) {

            logger.severe("Error sending rpc message: " + e.getMessage());
            controller.setFailed(e.getMessage());
            done.run(null);
        }


    }


    /**
     * Send rpc message to olad.
     *
     * @param msg RpcMessage
     * @throws Exception
     */
    private void sendMessage(RpcMessage msg) throws Exception {

        byte[] data = msg.toByteArray();

        int headerContent = (PROTOCOL_VERSION << 28) & VERSION_MASK;
        headerContent |= data.length & SIZE_MASK;
        byte[] header = ByteBuffer.allocate(4).order(ByteOrder.nativeOrder()).putInt(headerContent).array();

        if (logger.isLoggable(Level.FINEST)) {
            logger.info("Sending header " + header.length + " bytes");
            for (byte b : header) {
                System.out.format("0x%x ", b);
             }
            logger.info("Sending data " + data.length + " bytes");
            for (byte b : data) {
                 System.out.format("0x%x ", b);
              }
        }

        bos.write(header);
        bos.write(data);
        bos.flush();
    }


    /**
     * @return RpcMessage read back from olad.
     *
     * @throws Exception
     */
    private RpcMessage readMessage() throws Exception {

        byte[] header = new byte[4];
        bis.read(header);

        int headerValue = ByteBuffer.wrap(header).order(ByteOrder.nativeOrder()).getInt();
        int size = headerValue  & SIZE_MASK;

        byte[] data = new byte[size];
        bis.read(data);

        if (logger.isLoggable(Level.FINEST)) {
            logger.info("Received header ");
            for (byte b : header) {
                    System.out.format("0x%x ", b);
            }
            logger.info("Received data ");
            for (byte b : data) {
                    System.out.format("0x%x ", b);
            }
        }

        return RpcMessage.parseFrom(data);
    }
}

File: python/ola/rpc/StreamRpcChannel.py
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#
# StreamRpcChannel.py
# Copyright (C) 2005 Simon Newton

import binascii
import logging
import struct

from google.protobuf import service
from ola.rpc import Rpc_pb2
from ola.rpc.SimpleRpcController import SimpleRpcController

from ola import ola_logger

"""A RpcChannel that works over a TCP socket."""

__author__ = 'nomis52@gmail.com (Simon Newton)'


class OutstandingRequest(object):
  """These represent requests on the server side that haven't completed yet."""
  def __init__(self, request_id, controller):
    self.controller = controller
    self.id = request_id


class OutstandingResponse(object):
  """These represent requests on the client side that haven't completed yet."""
  def __init__(self, response_id, controller, callback, reply_class):
    self.callback = callback
    self.controller = controller
    self.id = response_id
    self.reply = None
    self.reply_class = reply_class


class StreamRpcChannel(service.RpcChannel):
  """Implements a RpcChannel over a TCP socket."""
  PROTOCOL_VERSION = 1
  VERSION_MASK = 0xf0000000
  SIZE_MASK = 0x0fffffff
  RECEIVE_BUFFER_SIZE = 8192

  def __init__(self, socket, service_impl, close_callback=None):
    """Create a new StreamRpcChannel.

    Args:
      socket: the socket to communicate on
      service: the service to route incoming requests to
    """
    self._service = service_impl
    self._socket = socket
    self._sequence = 0
    self._outstanding_requests = {}
    self._outstanding_responses = {}
    self._buffer = []  # The received data
    self._expected_size = None  # The size of the message we're receiving
    self._skip_message = False  # Skip the current message
    self._close_callback = close_callback
    self._log_msgs = False  # set to enable wire message logging
    if self._log_msgs:
      logging.basicConfig(level=logging.DEBUG)

  def SocketReady(self):
    """Read data from the socket and handle when we get a full message.

    Returns:
      True if the socket remains connected, False if it was closed.
    """
    data = self._socket.recv(self.RECEIVE_BUFFER_SIZE)
    if not data:
      ola_logger.info('OLAD Server Socket closed')
      if self._close_callback is not None:
        self._close_callback()
      return False

    self._buffer.append(data)
    self._ProcessIncomingData()
    return True

  def CallMethod(self, method, controller, request, response_pb, done):
    """Call a method.

    Don't use this directly, use the Stub object.

    Args:
      method: The MethodDescriptor to call
      controller: An RpcController object
      request: The request message
      response: The response class
      done: A closure to call once complete.
    """
    message = Rpc_pb2.RpcMessage()
    message.type = Rpc_pb2.REQUEST
    message.id = self._sequence
    message.name = method.name
    message.buffer = request.SerializeToString()
    self._SendMessage(message)
    self._sequence += 1

    if message.id in self._outstanding_responses:
      # fail any outstanding response with the same id, not the best approach
      # but it'll do for now.
      ola_logger.warning('Response %d already pending, failing now', message.id)
      response = self._outstanding_responses[message.id]
      response.controller.SetFailed('Duplicate request found')
      self._InvokeCallback(response)

    response = OutstandingResponse(message.id, controller, done, response_pb)
    self._outstanding_responses[message.id] = response

  def RequestComplete(self, request, response):
    """This is called on the server side when a request has completed.

    Args:
      request: the OutstandingRequest object that has completed.
      response: the response to send to the client
    """
    message = Rpc_pb2.RpcMessage()

    if request.controller.Failed():
      self._SendRequestFailed(request)
      return

    message.type = Rpc_pb2.RESPONSE
    message.id = request.id
    message.buffer = response.SerializeToString()
    self._SendMessage(message)
    del self._outstanding_requests[request.id]

  def _EncodeHeader(self, size):
    """Encode a version and size into a header.

    Args:
      size: the size of the rpc message

    Returns:
      The header field
    """
    header = (self.PROTOCOL_VERSION << 28) & self.VERSION_MASK
    header |= size & self.SIZE_MASK
    return struct.pack('=L', header)

  def _DecodeHeader(self, header):
    """Decode a header into the version and size.

    Args:
      header: the received header

    Return:
      A tuple in the form (version, size)
    """
    return ((header & self.VERSION_MASK) >> 28, header & self.SIZE_MASK)

  def _SendMessage(self, message):
    """Send an RpcMessage.

    Args:
      message: An RpcMessage object.

    Returns:
      True if the send succeeded, False otherwise.
    """
    data = message.SerializeToString()
    # combine into one buffer to send so we avoid sending two packets
    data = self._EncodeHeader(len(data)) + data
    # this log is useful for building mock regression tests
    if self._log_msgs:
      logging.debug("send->" + str(binascii.hexlify(data)))

    sent_bytes = self._socket.send(data)
    if sent_bytes != len(data):
      ola_logger.warning('Failed to send full datagram')
      return False
    return True

  def _SendRequestFailed(self, request):
    """Send a response indicating this request failed.

    Args:
      request: An OutstandingRequest object.

    """
    message = Rpc_pb2.RpcMessage()
    message.type = Rpc_pb2.RESPONSE_FAILED
    message.id = request.id
    message.buffer = request.controller.ErrorText()
    self._SendMessage(message)
    del self._outstanding_requests[request.id]

  def _SendNotImplemented(self, message_id):
    """Send a not-implemented response.

    Args:
      message_id: The message number.
    """
    message = Rpc_pb2.RpcMessage
    message.type = Rpc_pb2.RESPONSE_NOT_IMPLEMENTED
    message.id = message_id
    self._SendMessage(message)

  def _GrabData(self, size):
    """Fetch the next N bytes of data from the buffer.

    Args:
      size: the amount of data to fetch.

    Returns:
      The next size bytes of data, or None if there isn't enough.
    """
    data_size = sum([len(s) for s in self._buffer])
    if data_size < size:
      return None

    data = []
    size_left = size
    while size_left > 0:
      chunk = self._buffer.pop(0)
      if len(chunk) > size_left:
        # we only want part of it
        data.append(chunk[0:size_left])
        self._buffer.insert(0, chunk[size_left:])
        size_left = 0
      else:
        data.append(chunk)
        size_left -= len(chunk)

    return b''.join(data)

  def _ProcessIncomingData(self):
    """Process the received data."""

    while True:
      if not self._expected_size:
        # this is a new msg
        raw_header = self._GrabData(4)
        if not raw_header:
          # not enough data yet
          return
        if self._log_msgs:
          logging.debug("recvhdr<-" + str(binascii.hexlify(raw_header)))
        header = struct.unpack('=L', raw_header)[0]
        version, size = self._DecodeHeader(header)

        if version != self.PROTOCOL_VERSION:
          ola_logger.warning('Protocol mismatch: %d != %d', version,
                             self.PROTOCOL_VERSION)
          self._skip_message = True
        self._expected_size = size

      data = self._GrabData(self._expected_size)
      if not data:
        # not enough data yet
        return

      if self._log_msgs:
        logging.debug("recvmsg<-" + str(binascii.hexlify(data)))
      if not self._skip_message:
        self._HandleNewMessage(data)
      self._expected_size = 0
      self._skip_message = 0

  def _HandleNewMessage(self, data):
    """Handle a new Message.

    Args:
      data: A chunk of data representing a RpcMessage
    """
    message = Rpc_pb2.RpcMessage()
    message.ParseFromString(data)

    if message.type in self.MESSAGE_HANDLERS:
      self.MESSAGE_HANDLERS[message.type](self, message)
    else:
      ola_logger.warning('Not sure of message type %d', message.type())

  def _HandleRequest(self, message):
    """Handle a Request message.

    Args:
      message: The RpcMessage object.
    """
    if not self._service:
      ola_logger.warning('No service registered')
      return

    descriptor = self._service.GetDescriptor()
    method = descriptor.FindMethodByName(message.name)
    if not method:
      ola_logger.warning('Failed to get method descriptor for %s', message.name)
      self._SendNotImplemented(message.id)
      return

    request_pb = self._service.GetRequestClass(method)()
    request_pb.ParseFromString(message.buffer)
    controller = SimpleRpcController()
    request = OutstandingRequest(message.id, controller)

    if message.id in self._outstanding_requests:
      ola_logger.warning('Duplicate request for %d', message.id)
      self._SendRequestFailed(message.id)

    self._outstanding_requests[message.id] = request
    self._service.CallMethod(method, request.controller, request_pb,
                             lambda x: self.RequestComplete(request, x))

  def _HandleResponse(self, message):
    """Handle a Response message.

    Args:
      message: The RpcMessage object.
    """
    response = self._outstanding_responses.get(message.id, None)
    if response:
      response.reply = response.reply_class()
      response.reply.ParseFromString(message.buffer)
      self._InvokeCallback(response)

  def _HandleCanceledResponse(self, message):
    """Handle a Not Implemented message.

    Args:
      message: The RpcMessage object.
    """
    ola_logger.warning('Received a canceled response')
    response = self._outstanding_responses.get(message.id, None)
    if response:
      response.controller.SetFailed(message.buffer)
      self._InvokeCallback(response)

  def _HandleFailedReponse(self, message):
    """Handle a Failed Response message.

    Args:
      message: The RpcMessage object.
    """
    response = self._outstanding_responses.get(message.id, None)
    if response:
      response.controller.SetFailed(message.buffer)
      self._InvokeCallback(response)

  def _HandleNotImplemented(self, message):
    """Handle a Not Implemented message.

    Args:
      message: The RpcMessage object.
    """
    ola_logger.warning('Received a non-implemented response')
    response = self._outstanding_responses.get(message.id, None)
    if response:
      response.controller.SetFailed('Not Implemented')
      self._InvokeCallback(response)

  def _InvokeCallback(self, response):
    """Run the callback and delete the outstanding response.

    Args:
      response: the resoponse to run.
    """
    response.callback(response.controller, response.reply)
    del self._outstanding_responses[response.id]

  MESSAGE_HANDLERS = {
      Rpc_pb2.REQUEST: _HandleRequest,
      Rpc_pb2.RESPONSE: _HandleResponse,
      Rpc_pb2.RESPONSE_CANCEL: _HandleCanceledResponse,
      Rpc_pb2.RESPONSE_FAILED: _HandleFailedReponse,
      Rpc_pb2.RESPONSE_NOT_IMPLEMENTED: _HandleNotImplemented,
  }
